backend/src/Controllers/castVoteController.ts

```diff
@@ -0,0 +1,187 @@
+    if (validationErr){
+        const errMsg = "Invalid Ballot: "+ validationErr
+        Logger.info(req, errMsg);
+        throw new BadRequest(errMsg);
+    }
+
+    //TODO: currently we have both a value on the input Ballot, and the route param.
+    //do we want to keep both?  enforce that they match?
+    const targetElectionId = req.params.id;
+    inputBallot.election_id = targetElectionId;
+    const targetElection = await ElectionsModel.getElectionByID(targetElectionId, req);
+    if (targetElection == null){
+        const errMsg = "Invalid Ballot: invalid election Id";
+        Logger.info(req, errMsg);
+        throw new BadRequest(errMsg);
+    }
+
+    if (targetElection.state!=='open'){
+        Logger.info(req, "Ballot Rejected. Election not open.", targetElection);
+        throw new BadRequest("Election is not open");
+    }
+
+    const voterId = getVoterID(req, targetElection.settings.voter_id_type, user);
+    Logger.debug(req, "voterID = " + voterId);
+    const roll = await getOrCreateElectionRoll(targetElection, voterId, req);
+    assertVoterMayVote(targetElection, roll, req);
+
+
+    //some ballot info should be server-authorative
+    inputBallot.date_submitted = Date.now();
+    if (inputBallot.history == null){
+        inputBallot.history = [];
+    }
+    inputBallot.ballot_id = randomUUID();
+    //TODO, ensure the user ID is added to the ballot...
+    //should server-authenticate the user id based on auth token
+    inputBallot.history.push({
+        action_type:"submit",
+        actor: voterId,
+        timestamp:inputBallot.date_submitted,
+    });
+
+    Logger.debug(req, "Submit Ballot:", inputBallot);
+    if (roll != null){
+        roll.ballot_id = String(inputBallot.ballot_id);
+        roll.submitted = true;
+        if (roll.history == null){
+            roll.history = [];
+        }
+        roll.history.push({
+            action_type:"submit",
+            actor: voterId,
+            timestamp:inputBallot.date_submitted,
+        });
+    }
+
+    const savedBallot = await persistBallotToStore(inputBallot, roll, req);
+    res.status("200").json({ ballot: savedBallot} );
+};
+
+
+async function persistBallotToStore(ballot:Ballot, roll:ElectionRoll|null, ctx:ILoggingContext):Promise<Ballot> {
+    Logger.debug(ctx, "persisteBallotToStore");
+    var savedBallot;
+    try {
+        if (roll == null){
+            savedBallot = await BallotModel.submitBallot(ballot, ctx, `User submits a ballot`);
+        } else {
+            savedBallot = await CastVoteStore.submitBallot(ballot, roll, ctx, 'User Submits a Ballot');
+        }
+        if (!savedBallot){
+            throw new InternalServerError("Failed to cast Ballot");
+        }
+    } catch (err:any){
+        Logger.error(ctx, 'Failed to cast Ballot:  '+err.message);
+        throw err;
+    }
+
+    return savedBallot;
+}
+
+
+function getVoterID(req:IRequest, voterIdType:string, user:any):string {
```

Comment:
getVoterID, getOrCreateElectionRoll, and assertVoterMayVote could be used in other controllers where the voter is checking if they can vote. We might want to move them to a util file or export them.

Reply:
I'm leaning towards leaving them here for now (probably out of laziness) and then whenever they need to be used elsewhere, go ahead and make the move to contollerUtils?